File: js/packages/app/src/__tests__/features/object-references/external/polywrap/object-references-example/src/polywrap/external/global-functions/testReceiveReference.ts
import { concat, u32ToBuffer, IExternalWrapInstance, BaseTypeSerialization } from '@polywrap/reim-wrap-js';
import { WrapModule } from '../module/WrapModule';
import { WrapManifest } from '../../WrapManifest';
import { ExternalResource } from '../../dt/ExternalResource';
import { TestExternalClassWrapped } from "../../wrapped";

import { TestExternalClass } from "../../..";



export function testReceiveReference(
  arg: TestExternalClass,
): Promise<string> {
  return testReceiveReferenceFromInstance(
    WrapModule.wrapInstance,
    arg,
  );
};

export const create = (instance: IExternalWrapInstance) => {
  return (
    arg: TestExternalClass,
  ): Promise<string> => {
    return testReceiveReferenceFromInstance(
      instance, 
      arg,
    );
  };
};

export const testReceiveReferenceFromInstance = async (
  instance: IExternalWrapInstance | null, 
  arg: TestExternalClass,
): Promise<string> => {
  if (instance == null) {
    throw new Error("connect() or import() must be called before using this module");
  }

  const args = new TestReceiveReferenceArgs(
    arg,
  );

  const buffer = concat([
    u32ToBuffer(WrapManifest.External.GlobalFunction.TestReceiveReference),
    TestReceiveReferenceArgsWrapped.serialize(args),
  ]);

  const result = await instance.invokeResource(ExternalResource.InvokeGlobalFunction, buffer);

  
  return BaseTypeSerialization.deserialize<string>(result);
}

export class TestReceiveReferenceArgsWrapped {
  constructor(
    public arg: TestExternalClassWrapped,
  ) {
  }

  static serialize(value: TestReceiveReferenceArgs): Uint8Array {
    return new TextEncoder().encode(
      JSON.stringify(
        TestReceiveReferenceArgsWrapped.mapToSerializable(value)
      )
    );
  }

  static mapToSerializable(value: TestReceiveReferenceArgs): TestReceiveReferenceArgsWrapped {
    return new TestReceiveReferenceArgsWrapped(
      
      TestExternalClassWrapped.mapToSerializable(value.arg),
            
    );
  }
}

export class TestReceiveReferenceArgs {
  constructor(
    public arg: TestExternalClass,
  ) {
  }
}

File: rust/app/src/polywrap/external_references/src/polywrap/external/global-functions/testReceiveReference.rs
use reim_wrap::{ ExternalWrapModule };
use module::{ WrapManifest };
use dt::{ ExternalResource };
use external_module::{ WrapModule };
use wrapped::{ testReceiveReferenceWrapped };
use wrapped::{ TestExternalClassWrapped };

use crate::{ TestExternalClass };



pub async fn testReceiveReference(
  arg: TestExternalClass,
) -> string {
  testReceiveReferenceFromInstance(
    WrapModule.wrapInstance,
    arg,
  )
};

pub fn create(instance: dyn ExternalWrapModule) {
  |
    arg: TestExternalClass,
  | {
    testReceiveReferenceFromInstance(
      instance, 
      arg,
    )
  }
}

pub async fn testReceiveReferenceFromInstance (
  instance: Option<ExternalWrapModule>, 
  arg: TestExternalClass,
) -> string {
  if instance.is_none() {
    panic!("connect() or import() must be called before using this module");
  }

  let args = TestReceiveReferenceArgs.new(
    arg,
  );

  let buffer = [
    WrapManifest.External.GlobalFunction.TestReceiveReference.to_be_bytes(),
    TestReceiveReferenceArgsWrapped.serialize(args),
  ].concat();

  let result = instance.invokeResource(ExternalResource.InvokeGlobalFunction, buffer).await;

  
  BaseTypeSerialization.deserialize<string>(result)
}

struct TestReceiveReferenceArgsWrapped {
    pub arg: TestExternalClassWrapped,
}
impl TestReceiveReferenceArgsWrapped {
    pub fn new(
      arg: TestExternalClass,
    ) -> TestReceiveReferenceArgsWrapped {
      TestReceiveReferenceArgsWrapped {
        arg,
      }
    }

    pub fn serialize(value: TestReceiveReferenceArgs) -> Vec<u8> {
        json!(
            TestReceiveReferenceArgsWrapped.mapToSerializable(value)
        ).as_bytes()
    }

    pub fn mapToSerializable(value: TestReceiveReferenceArgs) -> TestReceiveReferenceArgsWrapped {
        TestReceiveReferenceArgsWrapped.new(
            
            TestExternalClassWrapped.mapToSerializable(value.arg),
                        
        )
    }
}

struct TestReceiveReferenceArgs {
    pub arg: TestExternalClass,
}

impl TestReceiveReferenceArgs {
  pub fn new(
    arg: TestExternalClass,
  ) -> TestReceiveReferenceArgs {
    TestReceiveReferenceArgs.new (
      arg,
    )
  }
}
